concurrency/ClientRouterProvider.java

```diff
@@ -0,0 +1,195 @@
+     * @param metricsFactory Metrics factory for creating metrics instances for each ClientRouter.
+     * @param clientEndpoint The endpoint for this router. Used when creating a ClientRouter.
+     * @param ruleEngineProvider The rules engine provider for each ClientRouter.
+     * @param timeSource Used to track time.
+     * @param requestRateReporter Tracks the statistics for a router.
+     */
+    ClientRouterProvider(MetricsFactory metricsFactory,
+                         EndpointsConfig endpointsConfig,
+                         Address clientEndpoint,
+                         RuleEngineProvider ruleEngineProvider,
+                         TimeSource timeSource,
+                         RequestRateReporter requestRateReporter,
+                         ServiceThrottling serviceThrottling,
+                         ClientThrottler clientThrottler) {
+        this.metricsFactory = metricsFactory;
+        this.clientEndpoint = clientEndpoint;
+        this.requestRateReporter = requestRateReporter;
+        this.timeSource = timeSource;
+        this.ruleEngineProvider = ruleEngineProvider;
+        this.serviceThrottling = serviceThrottling;
+        this.clientThrottler = clientThrottler;
+        this.endpointsConfig = endpointsConfig;
+
+        createRouters();
+    }
+
+    /**
+     * Use the EndpointsConfig object to create routers for all endpoints/aliases which are not already created.
+     */
+    public void createRouters() {
+        //TODO (SLR-1942): handle endpoint/alias deletion
+        for (Map.Entry<String, List<String>> aliases : endpointsConfig.getAliasesForLbs().entrySet()) {
+            String lbName = aliases.getKey();
+            ClientRouter clientRouter = clientRouterServiceMap.computeIfAbsent(lbName, this::buildClientRouter);
+            aliasMap.put(lbName, clientRouter);
+            for (String alias : aliases.getValue()) {
+                aliasMap.put(alias, clientRouter);
+            }
+        }
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    /**
+     * Gets the ClientRouter for a specific load balancer.
+     *      If the load balancer is not whitelisted then it will throw an AccessDeniedException.
+     *      If the load balancer does not have a client router mapped to it then a new client router will be created.
+     * @param loadBalancer The load balancer to look up.
+     * @return ClientRouter for the given loadBalancer.
+     * @throws AccessDeniedException is the load balancer is unrecognized
+     */
+    public ClientRouter getClientRouter(String loadBalancer) {
+        ClientRouter router = aliasMap.get(loadBalancer);
+        if (router == null) {
+            if (loadBalancer.startsWith(EndpointsConfig.SLR_CANARY_ENDPOINT_PREFIX)) {
+                router = clientRouterServiceMap.computeIfAbsent(loadBalancer, x ->
+                        buildClientRouter(loadBalancer));
+                aliasMap.put(loadBalancer, router);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `get()`, `null` check, and `put()` may not be thread-safe at lines: **95, 96, and 100**. Two threads can perform this **same check at the same time** and one thread can **overwrite** the value written by the other thread.

**Fix**  
 Consider replacing `put()` with `putIfAbsent()` to help prevent accidental overwriting. `putIfAbsent()` puts the value only if the `ConcurrentHashMap` does not contain the key and therefore avoids overwriting the value written there by the other thread&#39;s `putIfAbsent()`.

**More info**  
 The thread whose value was overwritten still has a reference to that value and returns that reference to its caller at line: **106**. `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.